pkg/controller/plan/adapter/ovirt/client.go

```diff
@@ -409,6 +418,107 @@ func (r *Client) cacert() []byte {
 	return nil
 }
 
-func (c *Client) Finalize(vms []*planapi.VMStatus, planName string) {
+func (r Client) Finalize(vms []*planapi.VMStatus, planName string) {
+	r.connect()
+	defer r.Close()
+	var wg sync.WaitGroup
+	wg.Add(len(vms))
+	for _, vm := range vms {
+		_, vmService, err := r.getVM(vm.Ref)
+		if err != nil {
+			return
+		}
+
+		go r.removePrecopies(vm.Warm.Precopies, vmService, &wg)
+	}
+
+	wg.Wait()
+}
+
+func (r Client) removePrecopies(precopies []planapi.Precopy, vmService *ovirtsdk.VmService, wg *sync.WaitGroup) {
+	if len(precopies) == 0 {
+		return
+	}
+
+	defer wg.Done()
+	snapsService := vmService.SnapshotsService()
+	for i := range precopies {
+		snapshotID := precopies[i].Snapshot
+		snapService := snapsService.SnapshotService(snapshotID)
+		correlationID := fmt.Sprintf("%s_finalize", snapshotID[0:8])
+		_, err := snapService.Remove().Query("correlation_id", correlationID).Send()
+		r.Log.Error(err, "Remove request for snapshot failed", "snapshotID", snapshotID)
+		if err != nil {
+			// Assume we could not remove because of an ongoing transfer
+			backoff := wait.Backoff{
+				Duration: time.Second * 3,
+				Factor:   0,
+				Jitter:   0,
+				Steps:    3,
+			}
+
+			err = wait.ExponentialBackoff(backoff, func() (bool, error) {
+				snapshotDisksResponse, err := snapService.DisksService().List().Send()
+				if err != nil {
+					r.Log.Error(err, "snapshotDiksService Request failed")
+					err = liberr.Wrap(err, "Disks not found")
+					return false, err
+				}
+				snapshotDisks, found := snapshotDisksResponse.Disks()
+				if !found {
+					r.Log.Error(err, "disks not found")
+					err = liberr.Wrap(err, "Disks not found")
+					return false, err
+				}
+				for _, disk := range snapshotDisks.Slice() {
+					if transferred, err := r.isDiskBeingTransferred(disk.MustId()); transferred || err != nil {
+						r.Log.Info("Disk is being transferred, retrying..")
+						return false, nil
+					} else {
+						// Try to remove the snapshot again
+						_, err = snapService.Remove().Query("correlation_id", correlationID).Send()
+						if err != nil {
+							err = liberr.Wrap(err)
+							return true, err
```

Comment:
returning true here in case of an error?

Reply:
if we couldn't get the snapshot it makes sense to stop waiting no? or I misunderstood?